mlir/lib/Dialect/Rock/Transforms/LowerRockReduce.cpp

```diff
@@ -0,0 +1,233 @@
+    return op.emitError("func op should not have a gridSize set for a "
+                        "standalone rock.reduce op");
+  }
+  funcOp->setAttr("block_size", adaptor.getBlockSizeAttr());
+  funcOp->setAttr("grid_size", adaptor.getGridSizeAttr());
+
+  auto threadViewVal =
+      createThreadView(adaptor.getIn(), blockSize, gridSize, loc, rewriter);
+  ArrayAttr sourceTransformsFromOp;
+  Value source;
+  std::tie(source, sourceTransformsFromOp) =
+      untransform(rewriter, threadViewVal);
+  auto loadOobDims = computeOobFromTransforms(rewriter, sourceTransformsFromOp);
+
+  auto threadViewShape = threadViewVal.getType().cast<ShapedType>().getShape();
+  SmallVector<int64_t> bounds(threadViewShape.size(), 1LL);
+  bounds[threadViewShape.size() - 1] =
+      threadViewShape[threadViewShape.size() - 1];
+  SmallVector<int64_t> strides(threadViewShape.size(), 1LL);
+
+  // Get current workgroup ID.
+  auto bid = rewriter.create<WorkgroupIdOp>(loc, rewriter.getIndexType());
+  // Get current workitem ID.
+  auto tid = rewriter.create<WorkitemIdOp>(loc, rewriter.getIndexType());
+  auto zeroConstantOp = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+  SmallVector<Value, 3> loadStartCoords = {bid, tid, zeroConstantOp};
+
+  SmallVector<Value> zeroes(threadViewShape.size(), zeroConstantOp);
+
+  auto outLoop = rewriter.create<TransformingForOp>(
+      loc, ArrayRef<ValueRange>{loadStartCoords},
+      ArrayRef<Attribute>{sourceTransformsFromOp}, ArrayRef<int64_t>(bounds),
+      ArrayRef<int64_t>(strides),
+      /*forceUnroll=*/true, /*useIndexDiffs=*/true);
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(outLoop.getBody());
+    auto loadCoords = outLoop.getLowerCoords(/*domain=*/0);
+    auto loadVal = rewriter.create<GlobalLoadOp>(
+        loc, adaptor.getIn().getType().cast<ShapedType>().getElementType(),
+        adaptor.getIn(), std::get<0>(loadOobDims), std::get<1>(loadOobDims),
+        loadCoords);
+    SmallVector<Value, 4> storeCoords;
+    for (const auto &idxAndVal : llvm::enumerate(loadCoords)) {
+      auto idx = idxAndVal.index();
+      auto val = idxAndVal.value();
+      if (idx == redAxis) {
+        storeCoords.push_back(zeroConstantOp);
+      } else {
+        storeCoords.push_back(val);
+      }
+    }
+    StoreMethod stMethod;
+    ReduceMethod rMethod = adaptor.getReduceMethod();
+    if (getStoreMethod(rMethod, stMethod).failed()) {
+      return op.emitError()
+             << "The Reduce Method" << getNameForReduceMethod(rMethod)
+             << " is not supported.!";
+    }
+    rewriter.create<BufferStoreOp>(
+        loc, loadVal, adaptor.getOut(), std::get<0>(loadOobDims),
+        std::get<1>(loadOobDims), storeCoords,
+        StoreMethodAttr::get(rewriter.getContext(), stMethod),
+        /*offset=*/IntegerAttr());
+  }
+  auto region = op->getParentRegion();
+  rewriter.eraseOp(op);
+  (void)runRegionDCE(rewriter, {*region});
```

Comment:
I don't think you need to do this?

Reply:
(Or, if you do, maybe it should be happening up in gridwise-gemm-to-blockwise as well)

Reply:
There are few dead-end transforms that created to get the maps for transformingFor.
Using this to remove them. 

I think gridwise-gemm-to-blockwise is calling a canonicalize as a cleanup pass -- somehow getting cleaned up there?

Reply:
Yeah, canonicalize does a DCE.

Also, you
1. Don't exactly _need_ to DCE those
2. Can pass an array of `TransformMap` attributes (or an `ArrayAttr` of them) as the third argument to `untransform()`

Reply:
Thanks for the tip on untransform ( I didn't know that) and done!